File: src/test/kotlin/school/charset/app/domain/exercise/StepTest.kt
package school.charset.app.domain.exercise

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe
import school.charset.app.domain.encoding.Encoding

class StepTest :
    FreeSpec({
        "Format" - {
            "valid step builds" {
                val step = Step.Format(choices = listOf("a", "b"), expected = "a")
                step.expected shouldBe "a"
            }

            "empty choices throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Format(choices = emptyList(), expected = "x")
                }
            }

            "expected not in choices throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Format(choices = listOf("a", "b"), expected = "c")
                }
            }
        }

        "Binary" - {
            "valid step builds" {
                Step.Binary(expected = "11101001", length = 8).length shouldBe 8
            }

            "zero length throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Binary(expected = "", length = 0)
                }
            }

            "negative length throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Binary(expected = "1", length = -1)
                }
            }

            "expected length mismatching length throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Binary(expected = "1010", length = 8)
                }
            }

            "expected with non-binary character throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Binary(expected = "1110A001", length = 8)
                }
            }
        }

        "BitGroups" - {
            "valid step builds" {
                Step.BitGroups(expected = listOf("00011", "101001")).expected.size shouldBe 2
            }

            "empty list throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.BitGroups(expected = emptyList())
                }
            }

            "empty group throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.BitGroups(expected = listOf("00011", ""))
                }
            }

            "non-binary character in any group throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.BitGroups(expected = listOf("00011", "10A001"))
                }
            }
        }

        "HexBytes" - {
            "valid step builds" {
                Step.HexBytes(expected = listOf(0xC3, 0xA9)).expected.size shouldBe 2
            }

            "empty list throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.HexBytes(expected = emptyList())
                }
            }

            "byte > 255 throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.HexBytes(expected = listOf(0xC3, 0x100))
                }
            }

            "byte < 0 throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.HexBytes(expected = listOf(-1, 0xA9))
                }
            }
        }

        "CodePointEntry" - {
            "valid code point builds" {
                Step.CodePointEntry(expected = 0x00E9).expected shouldBe 0xE9
            }

            "boundary U+0000 builds" {
                Step.CodePointEntry(expected = 0x0000).expected shouldBe 0
            }

            "boundary U+10FFFF builds" {
                Step.CodePointEntry(expected = 0x10FFFF).expected shouldBe 0x10FFFF
            }

            "negative throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.CodePointEntry(expected = -1)
                }
            }

            "above U+10FFFF throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.CodePointEntry(expected = 0x110000)
                }
            }

            "surrogate U+D800 throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.CodePointEntry(expected = 0xD800)
                }
            }

            "surrogate U+DFFF throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.CodePointEntry(expected = 0xDFFF)
                }
            }
        }

        "Endianness" - {
            "BigEndian builds" {
                Step.Endianness(expected = Encoding.Endian.BigEndian).expected shouldBe
                    Encoding.Endian.BigEndian
            }

            "LittleEndian builds" {
                Step.Endianness(expected = Encoding.Endian.LittleEndian).expected shouldBe
                    Encoding.Endian.LittleEndian
            }
        }

        "UsefulBitCount" - {
            "valid count builds" {
                Step.UsefulBitCount(expected = 11).expected shouldBe 11
            }

            "lower bound 1 builds" {
                Step.UsefulBitCount(expected = 1).expected shouldBe 1
            }

            "upper bound 32 builds" {
                Step.UsefulBitCount(expected = 32).expected shouldBe 32
            }

            "zero throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.UsefulBitCount(expected = 0)
                }
            }

            "negative throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.UsefulBitCount(expected = -1)
                }
            }

            "above 32 throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.UsefulBitCount(expected = 33)
                }
            }
        }

        "Offset" - {
            "valid value builds" {
                Step.Offset(expected = 0xF389).expected shouldBe 0xF389
            }

            "lower bound 0 builds" {
                Step.Offset(expected = 0).expected shouldBe 0
            }

            "upper bound 0xFFFFF builds" {
                Step.Offset(expected = 0xFFFFF).expected shouldBe 0xFFFFF
            }

            "negative throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Offset(expected = -1)
                }
            }

            "above 0xFFFFF throws" {
                shouldThrow<IllegalArgumentException> {
                    Step.Offset(expected = 0x100000)
                }
            }
        }
    })

File: test/charset/exercise/step_test.exs
defmodule Charset.Exercise.StepTest do
  use ExUnit.Case, async: true

  alias Charset.Exercise.Step

  describe "Format" do
    test "valid step builds" do
      step = Step.Format.new!(["a", "b"], "a")
      assert step.choices == ["a", "b"]
      assert step.expected == "a"
    end

    test "empty choices raises" do
      assert_raise ArgumentError, fn -> Step.Format.new!([], "x") end
    end

    test "expected not in choices raises" do
      assert_raise ArgumentError, fn -> Step.Format.new!(["a", "b"], "c") end
    end
  end

  describe "Binary" do
    test "valid step builds" do
      step = Step.Binary.new!("11101001", 8)
      assert step.expected == "11101001"
      assert step.length == 8
    end

    test "zero length raises" do
      assert_raise ArgumentError, fn -> Step.Binary.new!("", 0) end
    end

    test "negative length raises" do
      assert_raise ArgumentError, fn -> Step.Binary.new!("1", -1) end
    end

    test "expected length mismatching length raises" do
      assert_raise ArgumentError, fn -> Step.Binary.new!("1010", 8) end
    end

    test "expected with non-binary character raises" do
      assert_raise ArgumentError, fn -> Step.Binary.new!("1110A001", 8) end
    end
  end

  describe "BitGroups" do
    test "valid step builds" do
      step = Step.BitGroups.new!(["00011", "101001"])
      assert step.expected == ["00011", "101001"]
    end

    test "empty list raises" do
      assert_raise ArgumentError, fn -> Step.BitGroups.new!([]) end
    end

    test "empty group raises" do
      assert_raise ArgumentError, fn -> Step.BitGroups.new!(["00011", ""]) end
    end

    test "non-binary character in any group raises" do
      assert_raise ArgumentError, fn -> Step.BitGroups.new!(["00011", "10A001"]) end
    end
  end

  describe "HexBytes" do
    test "valid step builds" do
      step = Step.HexBytes.new!([0xC3, 0xA9])
      assert step.expected == [0xC3, 0xA9]
    end

    test "empty list raises" do
      # Built dynamically: a literal [] lets the type system prove the call
      # always fails and flag it - which is exactly what this test asserts.
      empty = Enum.drop([0xC3], 1)
      assert_raise ArgumentError, fn -> Step.HexBytes.new!(empty) end
    end

    test "byte > 255 raises" do
      assert_raise ArgumentError, fn -> Step.HexBytes.new!([0xC3, 0x100]) end
    end

    test "byte < 0 raises" do
      assert_raise ArgumentError, fn -> Step.HexBytes.new!([-1]) end
    end
  end

  describe "CodePointEntry" do
    test "valid code point builds" do
      assert Step.CodePointEntry.new!(0xE9).expected == 0xE9
    end

    test "boundary U+0000 builds" do
      assert Step.CodePointEntry.new!(0x0000).expected == 0x0000
    end

    test "boundary U+10FFFF builds" do
      assert Step.CodePointEntry.new!(0x10FFFF).expected == 0x10FFFF
    end

    test "negative raises" do
      assert_raise ArgumentError, fn -> Step.CodePointEntry.new!(-1) end
    end

    test "above U+10FFFF raises" do
      assert_raise ArgumentError, fn -> Step.CodePointEntry.new!(0x110000) end
    end

    test "surrogate U+D800 raises" do
      assert_raise ArgumentError, fn -> Step.CodePointEntry.new!(0xD800) end
    end

    test "surrogate U+DFFF raises" do
      assert_raise ArgumentError, fn -> Step.CodePointEntry.new!(0xDFFF) end
    end
  end

  describe "Endianness" do
    test "big endian builds" do
      assert Step.Endianness.new!(:big).expected == :big
    end

    test "little endian builds" do
      assert Step.Endianness.new!(:little).expected == :little
    end
  end

  describe "UsefulBitCount" do
    test "valid count builds" do
      assert Step.UsefulBitCount.new!(11).expected == 11
    end

    test "lower bound 1 builds" do
      assert Step.UsefulBitCount.new!(1).expected == 1
    end

    test "upper bound 32 builds" do
      assert Step.UsefulBitCount.new!(32).expected == 32
    end

    test "zero raises" do
      assert_raise ArgumentError, fn -> Step.UsefulBitCount.new!(0) end
    end

    test "negative raises" do
      assert_raise ArgumentError, fn -> Step.UsefulBitCount.new!(-1) end
    end

    test "above 32 raises" do
      assert_raise ArgumentError, fn -> Step.UsefulBitCount.new!(33) end
    end
  end

  describe "Offset" do
    test "valid value builds" do
      assert Step.Offset.new!(0xF389).expected == 0xF389
    end

    test "lower bound 0 builds" do
      assert Step.Offset.new!(0).expected == 0
    end

    test "upper bound 0xFFFFF builds" do
      assert Step.Offset.new!(0xFFFFF).expected == 0xFFFFF
    end

    test "negative raises" do
      assert_raise ArgumentError, fn -> Step.Offset.new!(-1) end
    end

    test "above 0xFFFFF raises" do
      assert_raise ArgumentError, fn -> Step.Offset.new!(0x100000) end
    end
  end

  describe "type_id/1" do
    test "every step type has its stable id" do
      assert Step.type_id(Step.Format.new!(["a"], "a")) == "format"
      assert Step.type_id(Step.Binary.new!("1", 1)) == "binary"
      assert Step.type_id(Step.BitGroups.new!(["1"])) == "bit-groups"
      assert Step.type_id(Step.HexBytes.new!([0])) == "hex-bytes"
      assert Step.type_id(Step.CodePointEntry.new!(0xE9)) == "code-point"
      assert Step.type_id(Step.UsefulBitCount.new!(8)) == "useful-bit-count"
      assert Step.type_id(Step.Endianness.new!(:big)) == "endianness"
      assert Step.type_id(Step.Offset.new!(0)) == "offset"
    end
  end
end
